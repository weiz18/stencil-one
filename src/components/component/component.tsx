import { Component, Prop, h, JSX, Host, Watch, Event, State, EventEmitter, Method } from '@stencil/core';

@Component({
  tag: 'my-component',
  styleUrl: 'component.css',
  shadow: true
})
export class MyComponent {
  /**
   * The first name
   */
  @Prop() first: string;

  /**
   * The middle name
   */
  @Prop() middle: string;

  /**
   * The last name
   */
  @Prop() last: string;

  @Prop() language: string;

  @Prop({mutable: true}) values: Array<string> = [];

  @Watch('language')
  async onLanguageChange() {
    if (this.language) {
      return fetch('./assets/i18n/' + this.language + '.json')
        .then(response => {
          if (!response.ok) {
            throw new Error('HTTP error ' + response.status);
          }
          return response.json();
        })
        .then(json => {
          console.log('JSON', json);
        });
    }
  }

  @State() buttonFace: string = 'Click Me!';
  @State() clicked: boolean;

  @Event() somethingChanged: EventEmitter<any>;

  private inputElement: HTMLInputElement;

  format () {
    return [this.first, this.middle, this.last].filter (e => !!e).join(' ');
  }

  onInputChanged = () => {
    let value = this.inputElement.value;
    this.somethingChanged.emit({input: value});
  }

  @Method()
  async updateFace(value: string): Promise<string> {
    this.buttonFace = value;
    return this.buttonFace.toUpperCase();
  }

  private getText(): string {
    return this.format();
  }

  onClicked(value: string) {
    console.log('value is: ', value);
    this.clicked = !this.clicked;
    this.somethingChanged.emit({button: true});
  }

  render(): JSX.Element {
    return (
      <Host>
        <div class={{ nice: true, clicked: this.clicked }} >
          <span>
            Hello, World! I'm {this.getText()}
          </span>
          {this.values.map((item) => {
            return <div class="item">
                <span>{item}</span>
            </div>
        })}

          <button onClick={() => this.onClicked('ABCD')}>
            {this.buttonFace}
          </button>
          <input
          onInput={this.onInputChanged}
          ref={el => this.inputElement = el as HTMLInputElement}
        />
          <slot />
        </div>
      </Host>

    );
  }
}
