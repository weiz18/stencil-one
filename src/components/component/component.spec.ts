import {newSpecPage} from '@stencil/core/testing';

import {MyComponent} from './component';

describe('component', () => {
  afterEach(() => {
    jest.restoreAllMocks();
  })

  describe('instance', () => {
    it ('Should format the name', () => {
      let cmp = new MyComponent();
      cmp.first = 'Donald';
      cmp.last = 'Duck';
      expect(cmp.format()).toEqual('Donald Duck');
    });      
  });
  
  describe('render', () => {
    
    it('Should render with serializing shadow dom', async() => {
      const {root} = await newSpecPage({
        components: [MyComponent],
        html: `<my-component first="Donald" last="Duck">Some Text</my-component>`
      });

      expect(root).toEqualHtml(`
        <my-component first="Donald" last="Duck">
          <mock:shadow-root>
            <div class="nice">
              <span>
                Hello, World! I'm Donald Duck
              </span>
              <button>
                Click Me!
              </button>
              <input>
              <slot></slot>
            </div>
          </mock:shadow-root>
          Some Text
        </my-component>
      `);
      expect(root).toMatchSnapshot();
      expect(root.shadowRoot).toBeTruthy();
      expect(root.shadowRoot.querySelector('button')).toBeTruthy();
      expect(root.querySelector('button')).toBeFalsy();
    });


    it('Should render without serializing shadow dom', async() => {
      const {root} = await newSpecPage({
        components: [MyComponent],
        html: `<my-component first="Donald" last="Duck">Some Text</my-component>`,
        supportsShadowDom: false
      });

      expect(root).toMatchSnapshot();
      expect(root.shadowRoot).toBeFalsy();
      expect(root.querySelector('button')).toBeTruthy();

    });

    it('Should render setting content later', async() => {
      const html = `<my-component first="Donald" last="Duck">Some Text</my-component>`;
      const page = await newSpecPage({components: [MyComponent]});
      await page.setContent(html);
      expect(page.doc.body).toMatchSnapshot();
    });
  });

  fit('Should emit', async() => {
    const {root, win} = await newSpecPage({components: [MyComponent], html: `<my-component first="John" last="Doe"></my-component>`});
    let button = root
      .shadowRoot
      .querySelector('button');
    let buttonClicked = jest.fn();
    win.addEventListener('buttonClicked', buttonClicked);
    button.click();
    expect(buttonClicked).toHaveBeenCalled;
  });

  it('Should run method', async() => {
    const page = await newSpecPage({components: [MyComponent], html: `<my-component first="John" last="Doe"></my-component>`});

    let button = page
      .root
      .shadowRoot
      .querySelector('button');
    expect(button.textContent).toEqual('Click Me!')
    let ret = await page
      .rootInstance
      .updateFace('New Click!');
    await page.waitForChanges();
    expect(ret).toEqual('NEW CLICK!');
    expect(button.textContent).toEqual('New Click!');
  });

});
