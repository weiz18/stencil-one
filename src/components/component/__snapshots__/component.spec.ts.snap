// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component render Should render setting content later 1`] = `
<body>
  <my-component first="Donald" last="Duck">
    <mock:shadow-root>
      <div class="nice">
        <span>
          Hello, World! I'm Donald Duck
        </span>
        <button>
          Click Me!
        </button>
        <input>
        <slot></slot>
      </div>
    </mock:shadow-root>
    Some Text
  </my-component>
</body>
`;

exports[`component render Should render with serializing shadow dom 1`] = `
<my-component first="Donald" last="Duck">
  <mock:shadow-root>
    <div class="nice">
      <span>
        Hello, World! I'm Donald Duck
      </span>
      <button>
        Click Me!
      </button>
      <input>
      <slot></slot>
    </div>
  </mock:shadow-root>
  Some Text
</my-component>
`;

exports[`component render Should render without serializing shadow dom 1`] = `
<my-component first="Donald" last="Duck">
  <!---->
  <div class="nice">
    <span>
      Hello, World! I'm Donald Duck
    </span>
    <button>
      Click Me!
    </button>
    <input>
    Some Text
  </div>
</my-component>
`;
